### PR TITLE
Update README-Scotch.md

### PR DESCRIPTION
**Pull request by DanGiles**
Adding more information on the `tleaf` files, the scotch strategy strings, some test case runs and future directions.

### CODE REVIEW

File: README-Scotch.md [hunk 1]
```diff
@@ -52,17 +53,39 @@ In the following examples it is assumed that one mpi rank is mapped to each Cosm
 | 4               | 32                  | tleaf 3 4 1000 2 10 4 5 |
 | 8               | 64                  | tleaf 3 8 1000 2 10 4 5 |
 
-The user needs to define this tleaf structure and save it as `target.tgt` in the directory they will run SWIFT from. Ongoing work focuses on automatically generating this target architecture upon run time. 
+The first index denotes the number of layers of the tleaf structure and the subsequent index pairs are the number of nodes in a layer and the cost to communicate between them. For example the numbers represented in the 64 MPI rank case (`tleaf 3 8 1000 2 10 4 5`) are as follows: reading right to left the `3` denotes the number of layers in the leaf structure, `8` denotes the number of vertices in the first layer (which corresponds to Cosma 8 compute nodes), `1000` the relative cost for internode communication, `2` deontes the number of vertices in the second layer (number of sockets on each cosma 8 node), `10` relative cost for intersocket communication, `4` denotes the number of vertices in the final layer (number of NUMA regions per socket on cosma 8) and finally `5` is the relative cost of intrasocket communication.
```
Comment (ilectra) [inline README-Scotch.md:56]:
```suggestion
The first index denotes the number of layers of the tleaf structure and the subsequent index pairs are the number of nodes in a layer and the cost to communicate between them. For example the numbers represented in the 64 MPI rank case (`tleaf 3 8 1000 2 10 4 5`) are as follows: reading left to right the `3` denotes the number of layers in the leaf structure, `8` denotes the number of vertices in the first layer (which corresponds to 8 compute nodes), `1000` the relative cost for internode communication, `2` denotes the number of vertices in the second layer (number of sockets on each node), `10` relative cost for intersocket communication, `4` denotes the number of vertices in the final layer (number of NUMA regions per socket on cosma8) and finally `5` is the relative cost of intrasocket communication.  
```

File: README-Scotch.md [hunk 2]
```diff
@@ -40,10 +40,11 @@ Follow the usual installation [instructions](https://gitlab.cosma.dur.ac.uk/swif
...
+In the following examples it is assumed that one mpi rank is mapped to each Cosma 8 NUMA region. This enforces that `cpus-per-task=16` is defined in the SLURM submission script. The Cosma 8 nodes consist of 8 NUMA regions per node, with 4 NUMA regions per socket. Example `tleaf` files for various setups are given below, where the intrasocket communication cost between NUMA regions is set at _5_, intranode but across sockets is set at _10_ and the internode cost is set at _1000_. These weightings are estimated values but have been shown to give satisfactory results in the testcases explored. An estimate of the relative latency between NUMA regions on a node can be obtained by using [hwloc](https://github.com/open-mpi/hwloc), specifically by using `hwloc-distances`.
 
+**Example tleaf graphs to represent various Cosma 8 configurations**
```
Comment (ilectra) [inline README-Scotch.md:47]:
```suggestion
**Example tleaf graphs to represent various Cosma8 configurations** 
```

File: README-Scotch.md [hunk 3]
```diff
@@ -40,10 +40,11 @@ Follow the usual installation [instructions](https://gitlab.cosma.dur.ac.uk/swif
...
+Scotch carries out a _mapping_ of a _source_ (or process) graph onto a _target_ (or architecture) graph. The weighted _source_ graph is generated by SWIFT and it captures the computation and communication cost across the computational domain. The _target_ graph defines the communication cost across the available computing architecture. Therefore, to make use of the Scotch _mapping_ alogrithms a target architecture file (_target.tgt_) must be generated and it should mirror the set up of the cluster being used. Scotch provides optimised architecture files which capture most HPC set ups. For Cosma 8 runs we will be targetting NUMA regions therefore we have modelled the architecture as a `tleaf` structure.
```
Comment (ilectra) [inline README-Scotch.md:43]:
```suggestion
Scotch carries out a _mapping_ of a _source_ (or process) graph onto a _target_ (or architecture) graph. The weighted _source_ graph is generated by SWIFT and it captures the computation and communication cost across the computational domain. The _target_ graph defines the communication cost across the available computing architecture. Therefore, to make use of the Scotch _mapping_ alogrithms a target architecture file (_target.tgt_) must be generated and it should mirror the set up of the cluster being used. Scotch provides optimised architecture files which capture most HPC set ups. For Cosma8 runs we will be targetting NUMA regions therefore we have modelled the architecture as a `tleaf` structure. 
```

File: README-Scotch.md [hunk 4]
```diff
@@ -52,17 +53,39 @@ In the following examples it is assumed that one mpi rank is mapped to each Cosm
...
+Scotch carries out the mapping using various strategies which are outlined in the [documentation](https://gitlab.inria.fr/scotch/scotch/-/blob/master/doc/scotch_user7.0.pdf), a list of the strategies trialed include: `SCOTCH_STRATDEFAULT`, `SCOTCH_STRATBALANCE`, `SCOTCH_STRATQUALITY` and `SCOTCH_STRATSPEED`. The Scotch strategy string is passed to the Mapping functions. For the runs carried out here it was found that the global flag `SCOTCH_STRATBALANCE` and an imbalance ratio of `0.05` worked best. These values are passed to [`SCOTCH_stratGraphMapBuild`](https://github.com/UCL/swiftsim/blob/cb06b0e5c3d8457c474d0084d973f437d29b20d8/src/partition.c#L1657). 
+
+One issue with Scotch is that when the number of mpi ranks is comparable to the dimensionality of the modelled SWIFT system the optimal mapping strategy doesn't neccessarily map to all available NUMA regions. At present this isn't handled explicity in the code and the paritition reverts to a vectorised or previous partitioning.
```
Comment (ilectra) [inline README-Scotch.md:67]:
```suggestion
One issue with Scotch is that when the number of mpi ranks is comparable to the dimensionality of the modelled SWIFT system the optimal mapping strategy doesn't necessarily map to all available NUMA regions. At present this isn't handled explicitly in the code and the partition reverts to a vectorised or previous partitioning.
```

Comment (ilectra) [reply README-Scotch.md:67]: What does "the partition reverts to a vectorised or previous partitioning." mean? Or would this be obvious if I read the Scotch docs more carefully? :flushed:

File: README-Scotch.md [hunk 5]
```diff
@@ -52,17 +53,39 @@ In the following examples it is assumed that one mpi rank is mapped to each Cosm
...
+
+The SWIFT edge and vertex weights are estimated in the code, however edge weights are not symmetric - this causes an issue with SWIFT. Therefore, in the SCOTCH Graph the edge weigths are updated to equal to the sum of the two associated edge weights as calculated from SWIFT.
+
+Test Runs
+----------------
+Teh following results were obtained on cosma 8 running with the `SCOTCH_STRATBALANCE` strategy string and an imbalance ratio of `0.05`.
```
Comment (ilectra) [inline README-Scotch.md:73]:
```suggestion
The following results were obtained on cosma8 running with the `SCOTCH_STRATBALANCE` strategy string and an imbalance ratio of `0.05`.
```

**Review by ilectra — COMMENTED** [review]
That's very useful, thanks! I made some typo corrections, and changed most of the "Cosma 8" mentions to "Cosma8" - I think that's the official name of the machine, and especially when you're discussing 8 nodes on "Cosma 8" it becomes really confusing :) . Please fix any that I missed!

File: README-Scotch.md (hunk 4 above)
Comment (DanGiles) [reply README-Scotch.md:67]: This is the fall back partitioning used by SWIFT, see [here](https://github.com/UCL/swiftsim/blob/be7d0f1ceef6de3a94017751f8fb8b58ff19614e/src/partition.c#L2491C2-L2493C62).

**Pull request merged**